File: 49주차/부대복귀.py
from collections import defaultdict,deque

def solution(n, roads, sources, dest):
    answer = []
    dict = defaultdict(list)
    for a,b in roads:
        dict[a].append(b)
        dict[b].append(a)
        
    INF = int(1e9)
    def bfs(start):
        q = deque()
        
        q.append((start,0))
        distance = [INF]*(n+1)
        distance[start] = 0
        while q:
            now,dist = q.popleft()
            if distance[now] < dist:
                continue
                
            for i in dict[now]:
                cost = dist+1
                if cost < distance[i]:
                    distance[i] = cost
                    q.append((i, cost))
                
        return distance           
    arr = bfs(dest)
    for v in sources:
        if arr[v] != INF:
            answer.append(arr[v])
        else:
            answer.append(-1)# 갈 수 없다면 -1.
    return answer
# 모든 거리가 1이므로 bfs로 대체 가능
# 여러 출발 지점에서 부터, 같은 도착지점까지의 거리를 알면 되므로, 
# 도착지점 중심의 다익스트라 구하기.
# 프로그래머스

File: 50주차/부대복귀.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}
class Queue {
  constructor() {
    this.length = 0;
    this.head = 0;
    this.rear = 0;
  }
  enqueue(data) {
    const node = new Node(data);
    if (!this.head) {
      this.head = node;
    } else {
      this.rear.next = node;
    }
    this.rear = node;
    this.length++;
  }
  dequeue() {
    if (!this.head) {
      return false;
    }
    const data = this.head.data;
    this.head = this.head.next;
    this.length--;
    return data;
  }
}
function solution(n, roads, sources, destination) {
  var answer = [];
  let g = new Array(n + 1).fill().map((e) => new Array());
  roads.forEach(([a, b]) => {
    g[a].push(b);
    g[b].push(a);
  });
  const time = new Array(n + 1).fill(Infinity);
  const da = (start) => {
    time[start] = 0;
    const q = new Queue();
    q.enqueue([start, 0]);
    while (q.length) {
      const [now, dist] = q.dequeue();
      if (time[now] < dist) continue;
      for (let v of g[now]) {
        let cost = dist + 1;
        if (cost < time[v]) {
          time[v] = cost;
          q.enqueue([v, cost]);
        }
      }
    }
  };

  da(destination);
  for (let v of sources) {
    time[v] === Infinity ? answer.push(-1) : answer.push(time[v]);
  }
  return answer;
}
// 다익스트라. -> 모두 길이가 1인 경우에서 최단거리 구하기
// js는 큐를 따로 구현해야함.
